### DISCUSSION TITLE
Disable opentelemetry-go based on a flag

### DISCUSSION THREAD

**Discussion by Ajaykumarkolipaka** [open-telemetry/opentelemetry-go#2659] [Q&A]
Hello Community,
Can someone know best way to tackle this?

I'm following this [doc](https://opentelemetry.io/docs/instrumentation/go/getting-started/) to instrument the code. I want to disable the tracer feature so that there won't be any span creation and exporter tasks will happen.

1. I tried to put **alwaysOffSampler**, by doing this spans are not getting exported but spans are getting created. How can I disable completely without make any code changes?
2. using **NewNoopTracerProvider** creats no ops spans, but will it has any performance impact?

**Comment by dmathieu** [comment]
OpenTelemetry will default all spans creation to noop unless another tracer provider is included.

For example, in the [named tracer example](https://github.com/open-telemetry/opentelemetry-go/blob/main/example/namedtracer/main.go), if you don't call the `initTracer` method, nothing will happen. But your code will keep running exactly the same.
So the best way to disable tracing in your app while keeping the otel code around is to not configure any tracer.

As for performance of the noop provider, the impact should be minimal. You can see how the provider and spans work there: https://github.com/open-telemetry/opentelemetry-go/blob/main/trace/noop.go

**Comment by Ajaykumarkolipaka** [comment]
Hi @dmathieu, Thanks for your suggestion.
I skipped the call to **initTracer** method based on enable/disable flag, but i'm running into null pointer exception when i'm making a call to tracer.Start(ctx, spanName) method as the tracer is not initialised. Any suggestions how to fix this?

**Reply by dmathieu** [reply]
Could you share the full error?

**Reply by Ajaykumarkolipaka** [reply]
`if enableTracer {
		shutdown := initTracer()
		defer shutdown()
	}`
`_, span := tracer.Start(ctx, "spanName")`

**Error log:**
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x98ca45]

we are not initialising the tracer but other part of the code is trying to create the spans using tracer object. as the tracer object is nil, it's throwing above error.

**Reply by dmathieu** [reply]
Please include a full stack trace, not just the error.

Even when the tracer isn't configured, you shouldn't get an invalid memory address, unless something is wrong. When not configured, the library defaults to a noop tracer.

As an example, the following code runs properly:

```golang
package main

import (
	"context"

	"go.opentelemetry.io/otel"
)

func main() {
	tracer := otel.Tracer("")

	ctx := context.Background()
	ctx, span := tracer.Start(ctx, "test")
	defer span.End()
}
```

**Reply by Ajaykumarkolipaka** [reply]
As I want to enable or disable the tracer based on a flag, if configured this way
if the flag is set to **false**: noop tracer will be initialised and sampling rate will be set to 0.

if I configured in above manner, can I say tracer got disabled? or else will there be any action happening in the background?

**Reply by Ajaykumarkolipaka** [reply]
I did above configuration and tried to log the traceId and spanId, I'm getting all 0's .
traceId=00000000000000000000000000000000, spanId=0000000000000000

**Comment by gmann42** [comment]
@Ajaykumarkolipaka Did you manage to solve this ? I am looking for exactly same thing.

**Comment by Hannarong98** [comment]
a little bit late, but try:

```go
// in func SetupOTelSDK
if(traceEnabled == "false") {
   otel.SetTracerProvider(noop.NewTracerProvider())
} else {
  // use non noop provider
}
```
